build.ps1

```diff
@@ -1,183 +1,72 @@
-    }
-    if (-not (Test-Path $markdownDocsPath -Verbose:$verbosity))
-    {
-        throw "Cannot find markdown documentation folder."
-    }
-    New-ExternalHelp -Path $markdownDocsPath -OutputPath $outputDocsPath -Force -Verbose:$verbosity
-}
-
-# Appveyor errors out due to $profile being null. Hence...
-$moduleRootPath = "$HOME/Documents/WindowsPowerShell/Modules"
-if ($null -ne $profile)
-{
-    $moduleRootPath = Join-Path (Split-Path $profile) 'Modules'
-}
-$modulePSSAPath = Join-Path $moduleRootPath 'PSScriptAnalyzer'
-if ($Install)
-{
-    if (-not (Test-Path $moduleRootPath))
-    {
-        New-Item -Path $moduleRootPath -ItemType Directory -Force -Verbose:$verbosity
-    }
-    if (-not (Test-Path -Path $destinationPath))
-    {
-        throw "Please build the module first."
-    }
-    Copy-Item -Path $destinationPath -Destination $modulePSSAPath -Recurse -Verbose:$verbosity
-}
-
-if ($Test)
-{
-    Import-Module -Name Pester -MinimumVersion 4.3.1 -ErrorAction Stop
-    Function GetTestRunnerScriptContent($testPath)
-    {
-        $x = @"
-        cd $testPath
-        Invoke-Pester
-"@
-        return $x
-    }
-
-    Function CreateTestRunnerScript($testPath)
-    {
-        $tmptmpFilePath = [System.IO.Path]::GetTempFileName()
-        $tmpFilePath = $tmptmpFilePath + '.ps1'
-        Move-Item $tmptmpFilePath $tmpFilePath -Verbose:$verbosity
-        $content = GetTestRunnerScriptContent $testPath
-        Set-Content -Path $tmpFilePath -Value $content -Verbose:$verbosity
-        return $tmpFilePath
-    }
-
-    Function GetTestPath($TestType)
-    {
-        if ($TestType -eq "engine")
-        {
-            $testPath = Join-Path $projectRoot "Tests/Engine"
-        }
-        else
-        {
-            $testPath = Join-Path $projectRoot "Tests/Rules"
+END {
+    Import-Module -Force (Join-Path $PSScriptRoot build.psm1)
+    if ( $Clean -or $Clobber ) {
+        Remove-Build
+        if ( $PSCmdlet.ParameterSetName -eq "Clean" ) {
+            return
         }
-        return $testPath
     }
 
-    Function RunTest($TestType, [Boolean] $DifferentProcess)
-    {
-        $testPath = GetTestPath($TestType)
-        if ($DifferentProcess)
-        {
-            $testScriptFilePath = CreateTestRunnerScript $testPath
-            Start-Process powershell -ArgumentList "-NoExit","-File $testScriptFilePath" -Verb runas
-            # clean up the test file
+    $setName = $PSCmdlet.ParameterSetName
+    switch ( $setName ) {
+        "BuildAll" {
+            Build-ScriptAnalyzer -All -Configuration $Configuration
         }
-        else
-        {
-            try
-            {
-                Push-Location .
-                ([scriptblock]::Create((GetTestRunnerScriptContent $testPath))).Invoke()
-            }
-            finally
-            {
-                Pop-Location
-
+        "BuildDoc" {
+            Build-ScriptAnalyzer -Documentation
+        }
+        "BuildOne" {
+            $buildArgs = @{
```

Comment:
Side note: I find I do this a lot in PowerShell as well (pass arguments down through commands). Perhaps it's an antipattern (Law of Demeter violation type of thing), but it would be nice to introduce a simpler way for commands to "inherit" all the common parameters between it and its caller (which haven't been overridden).

Reply:
i often create a hash earlier in the script and then modify it based on conditions.